src/operator/controllers/aws_iam/serviceaccount_old/serviceaccount_controller.go

```diff
@@ -0,0 +1,197 @@
+	"github.com/otterize/intents-operator/src/shared/operatorconfig"
+	"github.com/samber/lo"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/viper"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+type AWSRolePolicyManager interface {
+	DeleteOtterizeIAMRole(ctx context.Context, namespace string, name string) error
+	GenerateRoleARN(namespace string, name string) string
+	GetOtterizeRole(ctx context.Context, namespaceName, accountName string) (bool, *types.Role, error)
+	CreateOtterizeIAMRole(ctx context.Context, namespace string, name string, useSoftDeleteStrategy bool) (*types.Role, error)
+	CreateRolesAnywhereProfileForRole(ctx context.Context, role types.Role, namespace string, serviceAccountName string) (*rolesanywhereTypes.ProfileDetail, error)
+	DeleteRolesAnywhereProfileForServiceAccount(ctx context.Context, namespace string, serviceAccountName string) (bool, error)
+}
+
+type ServiceAccountReconciler struct {
+	client.Client
+	awsAgent                         AWSRolePolicyManager
+	markRolesAsUnusedInsteadOfDelete bool
+}
+
+func NewServiceAccountReconciler(client client.Client, awsAgent AWSRolePolicyManager, markRolesAsUnusedInsteadOfDelete bool) *ServiceAccountReconciler {
+	return &ServiceAccountReconciler{
+		Client:                           client,
+		awsAgent:                         awsAgent,
+		markRolesAsUnusedInsteadOfDelete: markRolesAsUnusedInsteadOfDelete,
+	}
+}
+
+func (r *ServiceAccountReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		WithOptions(controller.Options{RecoverPanic: lo.ToPtr(true)}).
+		For(&corev1.ServiceAccount{}).
+		Complete(r)
+}
+
+func (r *ServiceAccountReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := logrus.WithFields(logrus.Fields{"serviceAccount": req.Name, "namespace": req.Namespace})
+
+	serviceAccount := corev1.ServiceAccount{}
+
+	err := r.Get(ctx, req.NamespacedName, &serviceAccount)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, errors.Wrap(err)
+	}
+
+	isReferencedByPods, exists := getServiceAccountLabelValue(&serviceAccount)
+	if !exists {
+		logger.Debug("serviceAccount not labeled with credentials-operator.otterize.com/service-account, skipping")
+		return ctrl.Result{}, nil
+	}
+
+	isNoLongerReferencedByPodsOrIsBeingDeleted := serviceAccount.DeletionTimestamp != nil || !isReferencedByPods
+
+	if isNoLongerReferencedByPodsOrIsBeingDeleted {
+		err = r.awsAgent.DeleteOtterizeIAMRole(ctx, req.Namespace, req.Name)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to remove service account: %w", err)
+		}
+
+		if viper.GetBool(operatorconfig.EnableAWSRolesAnywhereKey) {
```

Comment:
Maybe it is worth aligning it with `markRolesAsUnusedInsteadOfDelete` (it doesn't matter if doing it with struct field or with viper, just use the same method for both flags)